tsfc/driver.py

```diff
@@ -153,43 +149,52 @@ def compile_integral(integral_data, form_data, prefix, parameters,
                              type(quad_rule))
 
         quadrature_multiindex = quad_rule.point_set.indices
-        quadrature_indices += quadrature_multiindex
+        quadrature_indices.extend(quadrature_multiindex)
 
         config = kernel_cfg.copy()
         config.update(quadrature_rule=quad_rule)
-        ir = fem.compile_ufl(integrand, interior_facet=interior_facet, **config)
-        if parameters["unroll_indexsum"]:
-            def predicate(index):
-                return index.extent <= parameters["unroll_indexsum"]
-            ir = opt.unroll_indexsum(ir, predicate=predicate)
-        ir = [gem.index_sum(expr, quadrature_multiindex) for expr in ir]
-        irs.append(ir)
-
-    # Sum the expressions that are part of the same restriction
-    ir = list(reduce(gem.Sum, e, gem.Zero()) for e in zip(*irs))
+        expressions = fem.compile_ufl(integrand,
+                                      interior_facet=interior_facet,
+                                      **config)
+        reps = mode.Integrals(expressions, quadrature_multiindex,
+                              argument_multiindices, params)
```

Comment:
It does not have to be GEM to GEM translation at this point yet, it only has to be GEM to GEM translation with `mode.Integrals` + `mode.flatten` combined. Modes may choose to do a GEM to GEM translation here (like `vanilla` here) and do a trivial concatenation later, or they may construct their representation here (perhaps just packing up these arguments here into an object), and then do the optimisation later at `mode.flatten` when all integrals are available. It is up to the mode to decide how to divide the work between `mode.Integrals` and `mode.flatten`.